File: 프로그래머스/2/389480. 완전범죄/완전범죄.java
import java.util.*;

class Solution {
    public int solution(int[][] info, int n, int m) {
        
        int len = info.length;
        
        boolean dp[][][] = new boolean[len+1][n][m];
        dp[0][0][0] = true;
        
        for(int k=0; k<len; k++){
            for(int i=0; i<n; i++){
                for(int j=0; j<m; j++){
                    if(!dp[k][i][j]) continue;
                    
                    int aSum = info[k][0];
                    int bSum = info[k][1];
                    
                    if(i + aSum < n) {
                        dp[k+1][i+aSum][j] = true;
                    }
                    
                    if(j + bSum < m){
                        dp[k+1][i][j+bSum] = true;
                    }                      
                }
            }
        }
        
        int min = Integer.MAX_VALUE;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(dp[len][i][j]){
                    min = Math.min(min, i);
                }
            }
        }
        
        if(min == Integer.MAX_VALUE){
            min = -1;
        }
               
        return min;
    }
}

File: 프로그래머스/2/389480. 완전범죄/완전범죄.js
function solution(info, n, m) {
    const len = info.length;
    const dp = Array.from({length : len + 1}, () => Array.from({length : n}, () => Array(m).fill(false)))
    
    dp[0][0][0] = true
    
    for(let k=0; k<len; k++){
        for(let i=0; i<n; i++){
            for(let j=0; j<m; j++){
                
                if(!dp[k][i][j]) continue;
                 
                let aSum = info[k][0]
                let bSum = info[k][1]
                
                if(aSum + i < n){
                    dp[k+1][aSum+i][j] = true;
                }
                
                if(bSum+j < m){
                    dp[k+1][i][bSum+j] = true;
                }
                
            }
        }
    }
                          
    var answer = Number.MAX_SAFE_INTEGER;
    
    for(let i=0; i<n; i++){
        for(let j=0; j<m; j++){
            if(!dp[len][i][j]) continue
            answer = Math.min(answer, i)
        }
    }
    
    if(answer == Number.MAX_SAFE_INTEGER){
        answer = -1
    }
    
    return answer;
}